File: types/dividend.py
# Unified Dividend Data Model
# This file defines the canonical dividend data structure used across the entire application

from typing import Optional, Dict, Any
from pydantic import BaseModel, Field, validator
from datetime import date, datetime
from decimal import Decimal
from enum import Enum


class DividendType(str, Enum):
    """Types of dividends"""
    CASH = "cash"
    STOCK = "stock"
    DRP = "drp"  # Dividend Reinvestment Plan


class DividendSource(str, Enum):
    """Sources of dividend data"""
    ALPHA_VANTAGE = "alpha_vantage"
    MANUAL = "manual"
    BROKER = "broker"


class DividendStatus(str, Enum):
    """Dividend confirmation status"""
    PENDING = "pending"
    CONFIRMED = "confirmed"
    EDITED = "edited"


class BaseDividendData(BaseModel):
    """
    Base dividend data model - used for global dividend reference table
    This represents the raw dividend data from external APIs
    """
    symbol: str = Field(..., description="Stock ticker symbol")
    ex_date: date = Field(..., description="Ex-dividend date")
    pay_date: date = Field(..., description="Payment date")
    amount_per_share: Decimal = Field(..., description="Dividend amount per share", ge=0)
    currency: str = Field(default="USD", description="Currency code")
    dividend_type: DividendType = Field(default=DividendType.CASH)
    source: DividendSource = Field(default=DividendSource.ALPHA_VANTAGE)
    
    # Optional dates
    declaration_date: Optional[date] = Field(None, description="Declaration date")
    record_date: Optional[date] = Field(None, description="Record date")
    
    @validator('symbol')
    def validate_symbol(cls, v):
        if not v or len(v) > 10:
            raise ValueError('Symbol must be 1-10 characters')
        return v.upper()
    
    @validator('currency')
    def validate_currency(cls, v):
        if len(v) != 3:
            raise ValueError('Currency must be 3-character code')
        return v.upper()


class UserDividendData(BaseDividendData):
    """
    User-specific dividend data model - extends base with user ownership information
    This is what gets returned to the frontend and represents the complete dividend record
    """
    id: str = Field(..., description="Unique dividend record ID")
    user_id: Optional[str] = Field(None, description="User ID (null for global dividends)")
    
    # User-specific ownership information
    shares_held_at_ex_date: Decimal = Field(..., description="Shares held at ex-dividend date", ge=0)
    current_holdings: Decimal = Field(..., description="Current shares held", ge=0)
    total_amount: Decimal = Field(..., description="Total dividend amount (per_share * shares_held)", ge=0)
    
    # Confirmation status based on transaction existence
    confirmed: bool = Field(..., description="True if user has confirmed/received dividend")
    status: DividendStatus = Field(default=DividendStatus.PENDING)
    
    # Additional user fields
    company: str = Field(..., description="Human-readable company name")
    notes: Optional[str] = Field(None, description="User notes")
    
    # Computed fields for frontend convenience
    is_future: bool = Field(default=False, description="True if pay_date is in the future")
    is_recent: bool = Field(default=False, description="True if dividend was added recently")
    
    # Metadata
    created_at: datetime = Field(..., description="Record creation timestamp")
    updated_at: Optional[datetime] = Field(None, description="Last update timestamp")
    
    @validator('total_amount', always=True)
    def calculate_total_amount(cls, v, values):
        """Ensure total_amount = amount_per_share * shares_held_at_ex_date"""
        if 'amount_per_share' in values and 'shares_held_at_ex_date' in values:
            calculated = values['amount_per_share'] * values['shares_held_at_ex_date']
            return calculated
        return v


class DividendSummary(BaseModel):
    """Dividend summary statistics for analytics"""
    total_received: Decimal = Field(default=0, description="Total dividends received (confirmed)")
    total_pending: Decimal = Field(default=0, description="Total pending dividends")
    ytd_received: Decimal = Field(default=0, description="Year-to-date received")
    confirmed_count: int = Field(default=0, description="Number of confirmed dividends")
    pending_count: int = Field(default=0, description="Number of pending dividends")


class DividendCreateRequest(BaseModel):
    """Request model for creating/updating dividend records"""
    user_id: str
    symbol: str
    ex_date: date
    pay_date: date
    amount_per_share: Decimal
    shares_held_at_ex_date: Decimal
    currency: str = "USD"
    dividend_type: DividendType = DividendType.CASH
    source: DividendSource = DividendSource.ALPHA_VANTAGE


class DividendConfirmRequest(BaseModel):
    """Request model for confirming dividend receipt"""
    dividend_id: str
    edited_amount: Optional[Decimal] = Field(None, description="Override total amount if edited")


class DividendResponse(BaseModel):
    """Standard API response wrapper for dividend operations"""
    success: bool
    data: Optional[UserDividendData] = None
    error: Optional[str] = None
    message: Optional[str] = None


class DividendListResponse(BaseModel):
    """Standard API response for dividend lists"""
    success: bool
    data: list[UserDividendData] = []
    metadata: Dict[str, Any] = {}
    total_count: int = 0
    error: Optional[str] = None


# Utility functions for data transformation

def global_dividend_to_user_dividend(
    global_dividend: BaseDividendData,
    user_id: str,
    shares_held: Decimal,
    current_holdings: Decimal,
    confirmed: bool,
    company_name: str,
    dividend_id: str,
    created_at: datetime,
    updated_at: Optional[datetime] = None
) -> UserDividendData:
    """Convert global dividend data to user-specific dividend data"""
    return UserDividendData(
        id=dividend_id,
        user_id=user_id,
        symbol=global_dividend.symbol,
        ex_date=global_dividend.ex_date,
        pay_date=global_dividend.pay_date,
        amount_per_share=global_dividend.amount_per_share,
        currency=global_dividend.currency,
        dividend_type=global_dividend.dividend_type,
        source=global_dividend.source,
        declaration_date=global_dividend.declaration_date,
        record_date=global_dividend.record_date,
        shares_held_at_ex_date=shares_held,
        current_holdings=current_holdings,
        total_amount=global_dividend.amount_per_share * shares_held,
        confirmed=confirmed,
        status=DividendStatus.CONFIRMED if confirmed else DividendStatus.PENDING,
        company=company_name,
        is_future=global_dividend.pay_date > date.today(),
        is_recent=(datetime.now() - created_at).days < 7,
        created_at=created_at,
        updated_at=updated_at
    )


def calculate_dividend_summary(dividends: list[UserDividendData]) -> DividendSummary:
    """Calculate summary statistics from dividend list"""
    current_year = datetime.now().year
    
    confirmed_dividends = [d for d in dividends if d.confirmed]
    pending_dividends = [d for d in dividends if not d.confirmed and d.shares_held_at_ex_date > 0]
    
    total_received = sum(d.total_amount for d in confirmed_dividends)
    total_pending = sum(d.total_amount for d in pending_dividends)
    ytd_received = sum(
        d.total_amount for d in confirmed_dividends 
        if d.pay_date.year == current_year
    )
    
    return DividendSummary(
        total_received=total_received,
        total_pending=total_pending,
        ytd_received=ytd_received,
        confirmed_count=len(confirmed_dividends),
        pending_count=len(pending_dividends)
    )

File: shared/models/dividend.ts
// Unified Dividend Data Model - Frontend TypeScript Types
// This mirrors the Python types for consistent data contracts

export type DividendType = 'cash' | 'stock' | 'drp';
export type DividendSource = 'alpha_vantage' | 'manual' | 'broker';
export type DividendStatus = 'pending' | 'confirmed' | 'edited';

/**
 * Base dividend data - represents raw dividend information from external APIs
 */
export interface BaseDividendData {
  symbol: string;
  ex_date: string; // ISO date string
  pay_date: string; // ISO date string
  amount_per_share: number;
  currency: string;
  dividend_type: DividendType;
  source: DividendSource;
  declaration_date?: string;
  record_date?: string;
}

/**
 * Complete user dividend record - what the frontend receives from the API
 * Contains all necessary information for display and actions
 */
export interface UserDividendData extends BaseDividendData {
  id: string;
  user_id?: string; // null for global dividends
  
  // User-specific ownership information (ALWAYS present in API responses)
  shares_held_at_ex_date: number;
  current_holdings: number;
  total_amount: number; // CALCULATED: amount_per_share * shares_held_at_ex_date
  
  // Confirmation status (based on transaction existence, not just a flag)
  confirmed: boolean;
  status: DividendStatus;
  
  // Display information
  company: string; // Human-readable company name
  notes?: string;
  
  // Computed convenience fields
  is_future: boolean;
  is_recent: boolean;
  
  // Metadata
  created_at: string; // ISO datetime string
  updated_at?: string;
}

/**
 * Dividend summary statistics for analytics dashboard
 */
export interface DividendSummary {
  total_received: number;
  total_pending: number;
  ytd_received: number;
  confirmed_count: number;
  pending_count: number;
}

/**
 * Request types for API calls
 */
export interface DividendConfirmRequest {
  dividend_id: string;
  edited_amount?: number; // Optional override for total amount
}

export interface DividendSyncRequest {
  symbol?: string; // If provided, sync only this symbol
}

/**
 * API Response wrappers
 */
export interface DividendResponse {
  success: boolean;
  data?: UserDividendData;
  error?: string;
  message?: string;
}

export interface DividendListResponse {
  success: boolean;
  data: UserDividendData[];
  metadata: {
    timestamp: string;
    user_id: string;
    confirmed_only: boolean;
    total_dividends: number;
    [key: string]: any;
  };
  total_count: number;
  error?: string;
}

export interface DividendSummaryResponse {
  success: boolean;
  data: {
    dividend_summary: DividendSummary;
    [key: string]: any;
  };
  error?: string;
}

/**
 * Utility types for frontend components
 */
export interface DividendTableRow {
  id: string;
  symbol: string;
  company: string;
  ex_date: string;
  pay_date: string;
  amount_per_share: number;
  total_amount: number;
  current_holdings: number;
  shares_held_at_ex_date: number;
  confirmed: string; // 'Yes' | 'No' for display
  currency: string;
  created_at: string;
  is_future: boolean;
  is_recent: boolean;
}

/**
 * Form validation types
 */
export interface DividendFormErrors {
  [field: string]: string;
}

/**
 * Utility functions for type conversions and calculations
 */

/**
 * Convert UserDividendData to DividendTableRow for ApexListView
 */
export function dividendToTableRow(dividend: UserDividendData): DividendTableRow {
  // Robust null/undefined handling for numeric fields
  const safeNumber = (value: any): number => {
    if (value === null || value === undefined || isNaN(Number(value))) {
      return 0;
    }
    return Number(value);
  };

  // DEBUG: Log the input dividend
  console.log('[dividendToTableRow] Input dividend:', {
    id: dividend.id,
    id_type: typeof dividend.id,
    symbol: dividend.symbol,
    total_amount: dividend.total_amount
  });

  const tableRow = {
    id: dividend.id || '',
    symbol: dividend.symbol || '',
    company: dividend.company || getCompanyDisplayName(dividend.symbol || ''),
    ex_date: dividend.ex_date || '',
    pay_date: dividend.pay_date || '',
    amount_per_share: safeNumber(dividend.amount_per_share),
    total_amount: safeNumber(dividend.total_amount),
    current_holdings: safeNumber(dividend.current_holdings),
    shares_held_at_ex_date: safeNumber(dividend.shares_held_at_ex_date),
    confirmed: dividend.confirmed ? 'Yes' : 'No',
    currency: dividend.currency || 'USD',
    created_at: dividend.created_at || '',
    is_future: dividend.is_future || false,
    is_recent: dividend.is_recent || false
  };

  // DEBUG: Log the output table row
  console.log('[dividendToTableRow] Output tableRow:', {
    id: tableRow.id,
    id_type: typeof tableRow.id,
    symbol: tableRow.symbol,
    total_amount: tableRow.total_amount
  });

  return tableRow;
}

/**
 * Format currency with proper symbol and precision
 */
export function formatDividendCurrency(amount: number | null | undefined, currency: string = 'USD'): string {
  // Handle null/undefined/NaN values safely
  const safeAmount = (amount === null || amount === undefined || isNaN(Number(amount))) ? 0 : Number(amount);
  
  return new Intl.NumberFormat('en-US', {
    style: 'currency',
    currency: currency,
    minimumFractionDigits: 2,
    maximumFractionDigits: 4
  }).format(safeAmount);
}

/**
 * Format date for dividend display
 */
export function formatDividendDate(dateString: string): string {
  return new Date(dateString).toLocaleDateString('en-US', {
    year: 'numeric',
    month: 'short',
    day: 'numeric'
  });
}

/**
 * Validate dividend data before sending to API
 */
export function validateDividendData(data: Partial<UserDividendData>): DividendFormErrors {
  const errors: DividendFormErrors = {};
  
  if (!data.symbol || data.symbol.length === 0) {
    errors.symbol = 'Symbol is required';
  }
  
  if (!data.ex_date) {
    errors.ex_date = 'Ex-date is required';
  }
  
  if (!data.pay_date) {
    errors.pay_date = 'Pay date is required';
  }
  
  if (!data.amount_per_share || data.amount_per_share <= 0) {
    errors.amount_per_share = 'Amount per share must be positive';
  }
  
  if (!data.shares_held_at_ex_date || data.shares_held_at_ex_date <= 0) {
    errors.shares_held_at_ex_date = 'Shares held must be positive';
  }
  
  return errors;
}

/**
 * Calculate total amount from per-share amount and shares
 */
export function calculateTotalDividendAmount(
  amountPerShare: number, 
  sharesHeld: number
): number {
  return amountPerShare * sharesHeld;
}

/**
 * Check if dividend is eligible for confirmation
 * (user held shares at ex-date and not already confirmed)
 */
export function isDividendConfirmable(dividend: UserDividendData): boolean {
  return !dividend.confirmed && dividend.shares_held_at_ex_date > 0;
}

/**
 * Get company name with fallback
 */
export function getCompanyDisplayName(symbol: string, company?: string): string {
  if (company && company.trim().length > 0) {
    return company;
  }
  
  // Fallback to symbol with "Corporation" suffix
  return `${symbol} Corporation`;
}